src/protocol/CheckpointTracker.sol

```diff
@@ -0,0 +1,85 @@
+    Checkpoint public provenCheckpoint;
+
+    /// @notice Verified transitions between two checkpoints
+    /// @dev the startCheckpoint is not necessarily valid, but the endCheckpoint is correctly built on top of it.
+    mapping(bytes32 startCheckpointHash => bytes32 endCheckpointHash) public transitions;
+
+    IPublicationFeed public immutable publicationFeed;
+
+    // This would usually be retrieved dynamically as in the current Taiko implementation, but for simplicity we are
+    // just setting it in the constructor
+    IVerifier public immutable verifier;
+
+    /// @param genesis the checkpoint commitment describing the initial state of the rollup
+    /// @param _publicationFeed the input data source that updates the state of this rollup
+    /// @param _verifier a contract that can verify the validity of a transition from one checkpoint to another
+    constructor(bytes32 genesis, address _publicationFeed, address _verifier) {
+        // set the genesis checkpoint commitment of the rollup - genesis is trusted to be correct
+        require(genesis != 0, "genesis checkpoint commitment cannot be 0");
+        provenCheckpoint.commitment = genesis;
+        publicationFeed = IPublicationFeed(_publicationFeed);
+        verifier = IVerifier(_verifier);
+    }
+
+    /// @notice Verifies a transition between two checkpoints. Update the latest `provenCheckpoint` if possible
+    /// @param start The initial checkpoint before the transition
+    /// @param end The final checkpoint after the transition
+    /// @param proof Arbitrary data passed to the `verifier` contract to confirm the transition validity
+    function proveTransition(Checkpoint calldata start, Checkpoint calldata end, bytes calldata proof) external {
+        bytes32 startCheckpointHash = keccak256(abi.encode(start));
+        bytes32 endCheckpointHash = keccak256(abi.encode(end));
+
+        require(end.commitment != 0, "Checkpoint commitment cannot be 0");
+        // TODO: once the proving incentive mechanism is in place we should reconsider this requirement because
+        // ideally we would use the proof that creates the longest chain of proven publications.
+        require(transitions[startCheckpointHash] == 0, "Checkpoint already has valid transition");
+        require(start.publicationId < end.publicationId, "Start must be before end");
+        require(end.publicationId < publicationFeed.getNextPublicationId(), "Publication does not exist");
+
+        verifier.verifyProof(
+            publicationFeed.getPublicationHash(start.publicationId),
+            publicationFeed.getPublicationHash(end.publicationId),
+            start.commitment,
+            end.commitment,
+            proof
+        );
+
+        if (start.publicationId == provenCheckpoint.publicationId && start.commitment == provenCheckpoint.commitment) {
+            provenCheckpoint = end;
+            emit CheckpointProven(end.publicationId, end.commitment);
```

Comment:
Shouldn't we reuse the same CheckpointProven event with the 4 arguments?

Reply:
Do you mean the `TransitionProven` event? We are emitting `TransitionProven` events whenever a valid transition is recorded, whether or not it eventually connects with the actual state changes. `CheckpointProven` implies that it's part of the correct chain.